### PR TITLE
Add ExtensionList#lookupFirst convenience method

### PR DESCRIPTION
**Pull request by daniel-beck**
I find myself typing `ExtensionList.lookup(Whatever.class).get(Whatever.class)` a lot to get the singleton instance for a registered `@Extension`.

If I'm not the only one (and haven't missed something obvious), maybe this is useful? @jenkinsci/code-reviewers More to tell me there's an obvious alternative I've missed.

### CODE REVIEW

File: core/src/main/java/hudson/ExtensionList.java [hunk 1]
```diff
@@ -418,6 +418,20 @@ private void fireOnChangeListeners() {
     }
 
     /**
+     * Convenience method allowing lookup of the first instance of a given type.
+     * Equivalent to {@code ExtensionList.lookup(Class).get(Class)}.
+     *
+     * @param type The type to look up.
+     * @param <U>
+     * @return
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/ExtensionList.java]: Empty param/return should be removed or documented

**Review by oleg-nenashev — COMMENTED** [review]
Yes, it is something we really need. Probably makes sense to refactor bits within the core.

A more strict method would be also useful:

```java
public static @CheckForNull <U,T> T lookup(Class<U> extensionType, Class<T> implementationType) {
        return lookup(extensionType).get(implementationType);
 }
```

**Comment by daniel-beck** [comment]
> more strict method

But what would you even gain from that? This one simplifies things. Your proposal saves ~4 characters?

**Comment by oleg-nenashev** [comment]
> But what would you even gain from that? This one simplifies things by getting rid of redundancy. Your proposal saves ~4 characters?

It also avoids the "YOLO" mode of `lookupFirst ()`. But probably you could just remove "first" from the method name, because the call guarantees to have only one result IIUC

**Comment by daniel-beck** [comment]
> because the call guarantees to have only one result IIUC

No, as I may pass e.g. `Builder.class`.

**Comment by oleg-nenashev** [comment]
`Builder`... Uh oh, it should not be stored in the extension implementation list imho. But if it is true now, then the current method naming is correct of course

**Review by oleg-nenashev — APPROVED** [review]

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
Javadoc still needs to be fixed

**Comment by daniel-beck** [comment]
@oleg-nenashev 

> it should not be stored in the extension implementation list imho

Good point, then whatever the Descriptor type for that is, if any. The point is, the list can be any `@Extension` type, including super types, from a single implementation to e.g. `Descriptor`. To get a singleton instance, you may `lookup(Super).get(Specific)` or just `lookup(Specific).get(Specific)`.

Notably, `lookup(Super).get(Super)` already works and just gets whatever entry first fits.

**Comment by daniel-beck** [comment]
@oleg-nenashev Done

**Comment by daniel-beck** [comment]
This is the kind of change where I expect @jglick to tell me it's stupid and redundant due to something I didn't know about, so asking him for review.

**Review by amuniz — APPROVED** [review]
:+1:

**Review by jglick — COMMENTED** [review]
Needs some sample calls, for example of `Uptime` from `AbstractProject._poll`.

File: core/src/main/java/hudson/ExtensionList.java [hunk 2]
```diff
@@ -418,6 +418,19 @@ private void fireOnChangeListeners() {
...
     /**
+     * Convenience method allowing lookup of the first instance of a given type.
+     * Equivalent to {@code ExtensionList.lookup(Class).get(Class)}.
+     *
+     * @param type The type to look up.
+     * @return the first instance of the given type in its list.
+     *
+     * @since TODO
+     */
+    public static @CheckForNull <U> U lookupFirst(Class<U> type) {
```
Comment (jglick) [inline core/src/main/java/hudson/ExtensionList.java]: `@CheckForNull` here more or less defeats the real need for this, which is for cases where you expect that there *must* be one instance. I would rather say something like

```java
ExtensionList<U> all = lookup(type);
if (all.size() != 1) throw new IllegalStateException(…);
return all.get(0);
```

And name it something like `lookUpSingleton`.

See for example `GlobalConfigurationCategory.get`, `Jenkins.doConfigExecutorsSubmit` call to `MasterBuildConfiguration`, `JenkinsLocationConfiguration.get`, `UserDetailsCache.get`, ad nauseam. Basically just Find Usages on `ExtensionList.get(Class)`. If you know that the singleton has been registered, then you expect to be able to load it without null checks. Maybe existing calls in fact just assume that the result is not null, meaning FindBugs would complain if we actually let it.

**Review by jglick — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/ExtensionList.java [hunk 3]
```diff
@@ -418,6 +418,24 @@ private void fireOnChangeListeners() {
...
+     *
+     * @param type The type to look up.
+     * @return the singleton instance of the given type in its list.
+     *
+     * @since TODO
+     */
+    public static @Nonnull <U> U lookupSingleton(Class<U> type) {
+        ExtensionList<U> all = lookup(type);
+        if (all.size() != 1) {
+            throw new IllegalStateException("Expected 1 instance of " + type.getName() + " but got " + all.size());
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/ExtensionList.java]: Needs to be Javadoced. `throws` would be also useful.
In the case of this method, I agree that `RuntimeException` is a right approach.

**Comment by amuniz** [comment]
`Javadoced` :smile: Love English language.

File: core/src/main/java/hudson/ExtensionList.java (hunk 3 above)
Comment (amuniz) [reply core/src/main/java/hudson/ExtensionList.java]: `Javadoced` :smile: Love English language.

Comment (daniel-beck) [reply core/src/main/java/hudson/ExtensionList.java]: @oleg-nenashev It's mentioned in https://github.com/jenkinsci/jenkins/pull/3021/files#diff-e3f986a0854b9b724f4e12690cdeeac3R423 and I see no point in adding it to the signature. It's a runtime exception after all.

**Comment by oleg-nenashev** [comment]
>  I see no point in adding it to the signature. It's a runtime exception after all.

@daniel-beck The only point of it is to have just another place where the API users are warned about that in IDE. Nothing else

**Review by oleg-nenashev — APPROVED** [review]
OK though `@throws` tag in Javadoc exits for a reason. 🐝

File: core/src/main/java/hudson/ExtensionList.java (hunk 3 above)
Comment (jglick) [reply core/src/main/java/hudson/ExtensionList.java]: Still a good practice; helpful for Javadoc viewers.

```java
/**
 * …
 * @throws IllegalStateException if there are no instances, or more than one
 */
```

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 4]
```diff
@@ -3709,9 +3709,8 @@ public synchronized void doConfigExecutorsSubmit( StaplerRequest req, StaplerRes
         try {
             JSONObject json = req.getSubmittedForm();
 
-            MasterBuildConfiguration mbc = MasterBuildConfiguration.all().get(MasterBuildConfiguration.class);
-            if (mbc!=null)
-                mbc.configure(req,json);
+            MasterBuildConfiguration mbc = ExtensionList.lookupSingleton(MasterBuildConfiguration.class);
+            mbc.configure(req,json);
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: or even inline var

**Review by jglick — APPROVED** [review]

**Comment by daniel-beck** [comment]
Addressed @jglick's review comments (including one identical to one by @oleg-nenashev ), so I think this is ready to go into the next weekly.

**Review by oleg-nenashev — APPROVED** [review]
🐝 Ready to go.

**Comment by oleg-nenashev** [comment]
retriggering CI

**Comment by oleg-nenashev** [comment]
@daniel-beck Sorry, I've forgotten to reopen it :( Was in my TODO list for the next weekly anyway

**Review by jglick — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
CI hates us :(

**Comment by oleg-nenashev** [comment]
🚢 🇮🇹

**Pull request merged**